### ISSUE TITLE
_two_factor_login view should be public to allow errors management on token send 

### ISSUE THREAD

**Issue by callejerog** [jwag956/flask-security#266]
To manage errors during token send it is necessary to change a "non public" method. Also making send_security_token returning the send result, it has to be managed in that method

**Comment by jwag956** [comment]
Let me rephrase - send_security_token should be intercept-able so that errors can be managed, AND it should be able to return error messages and the calling view handle them.

With a quick glance - since it takes a 'user' argument - it should be simple to add a user.send_security_token(xxx) which in turn calls the current send_security_token. That user method could be redefined in your User model implementation.

Does that sound like a good solution?

**Comment by callejerog** [comment]
First of all, sorry for my bad English.
The rephrase obviously sounds good. It is what i meant.

I'm not sure about the solution proposed. The call to send_security_token is into the method _two_factor_login, that does not return anything to the login view. 

As temporary solution I did:

- redefine login view to call my two_factor_login method
- in two_factor_login manage errors on send getting result from send_security_token and returning login form (available as argument) with proper message
- in send_security_token return result and manage fall-back, retries etc.

Please advise me if I did something wrong, even if conceptually.

**Comment by jwag956** [comment]
That's pretty much what I am proposing to do. 

Add a new method user.tf_send_security_token(**kwargs) - which two_factor_login() will call, and can return either None (everything ok) or an error message - which will be propagated back to the form. So - precisely what you did.

You would then be able to define tf_send_security_token() in your User model and perform any error handling you want.

I will try to get that into 3.4 in the next week or so...

**Comment by callejerog** [comment]
Great! Thanks.
